x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java

```diff
@@ -195,25 +199,56 @@ public void createResponse(String docId,
     /**
      * Stores the final response if the place-holder document is still present (update).
      */
-    public void updateResponse(String docId,
-                            Map<String, List<String>> responseHeaders,
-                            R response,
-                            ActionListener<UpdateResponse> listener) {
+    public void updateResponse(AsyncExecutionId executionId,
+                               Map<String, List<String>> responseHeaders,
+                               R response,
+                               ActionListener<IndexResponse> listener) {
         try {
-            Map<String, Object> source = new HashMap<>();
-            source.put(RESPONSE_HEADERS_FIELD, responseHeaders);
-            source.put(RESULT_FIELD, encodeResponse(response));
-            UpdateRequest request = new UpdateRequest()
-                .index(index)
-                .id(docId)
-                .doc(source, XContentType.JSON)
-                .retryOnConflict(5);
-            clientWithOrigin.update(request, listener);
+            final String encodedResponse = encodeResponse(response);
+            doUpdateResponse(executionId, responseHeaders, encodedResponse, 0, listener);
         } catch(Exception e) {
             listener.onFailure(e);
         }
     }
 
+    private void doUpdateResponse(AsyncExecutionId executionId, Map<String, List<String>> responseHeaders,
+                                  String encodedResponse, int retryCount, ActionListener<IndexResponse> listener) {
+        // Since the encodedResponse can be very large, we perform the update logic on the client to minimize
+        // the memory usage by UpdateRequest and UpdateHelper.
+        assert retryCount <= 5 : retryCount;
+        final String docId = executionId.getDocId();
+        final GetRequest getRequest = new GetRequest(index)
+            .id(docId)
+            .preference(executionId.getEncoded())
+            .realtime(true);
```

Comment:
should we disable source fetching for this GET request:
`.fetchSourceContext(DO_NOT_FETCH_SOURCE)`?